### ISSUE TITLE
Massive and fast handle leak in WCF stress scenarios

### ISSUE THREAD

**Issue by KKhurin** [dotnet/corefx#2793]
0:102> !handle 0 0 
Type           	Count
Event          	26762
Key            	20745

Here are the characteristic call stacks:

~~~
Handle = 0x000000000002ea84 - OPEN
Thread ID = 0x0000000000002d74, Process ID = 0x0000000000004528

0x00007ff8c74c225a: ntdll!ZwOpenKeyEx+0x000000000000000a
0x00007ff8c47c3799: KERNELBASE!LocalBaseRegOpenKey+0x00000000000001fa
0x00007ff8c47cf532: KERNELBASE!RegOpenKeyExInternalA+0x0000000000000132
0x00007ff8c47cf5e9: KERNELBASE!RegOpenKeyExA+0x0000000000000019
0x00007ff8c0b82b1c: winhttp!OpenIeKey+0x000000000000005c
0x00007ff8c0b82c08: winhttp!OpenProxySettingsKeysForRead+0x0000000000000080
0x00007ff8c0b998bb: winhttp!INTERNET_SESSION_HANDLE_OBJECT::LoadAutomaticProxyResolvers+0x000000000000019b
0x00007ff8c0b730e5: winhttp!INTERNET_SESSION_HANDLE_OBJECT::SetProxySettings+0x0000000000044ed5
0x00007ff8c0b2d95a: winhttp!WinHttpSetOptionInternal+0x0000000000000397
0x00007ff8c0b2e9e6: winhttp!WinHttpOpen+0x000000000000027a
0x00007ff841eeccbd: System_Net_Http!DomainBoundILStubClass.IL_STUB_PInvoke(IntPtr, UInt32, System.String, System.String, Int32)+0x000000000000010d
0x00007ff841eec8b3: System_Net_Http!System.Net.Http.WinHttpHandler.EnsureSessionHandleExists(RequestState)+0x00000000000000e3
0x00007ff841eebba1: System_Net_Http!System.Net.Http.WinHttpHandler+<StartRequest>d__1.MoveNext()+0x00000000000001f1

--------------------------------------
Handle = 0x000000000002dd10 - OPEN
Thread ID = 0x00000000000027b4, Process ID = 0x0000000000004528

0x00007ff8c74c165a: ntdll!ZwCreateEvent+0x000000000000000a
0x00007ff8c47c2ce8: KERNELBASE!CreateEventW+0x0000000000000084
0x00007ff8c0b2ed38: winhttp!INTERNET_SESSION_HANDLE_OBJECT::INTERNET_SESSION_HANDLE_OBJECT+0x00000000000002be
0x00007ff8c0b2e963: winhttp!WinHttpOpen+0x00000000000001f3
0x00007ff8c0bb9afb: winhttp!CreateWinHttpSession+0x0000000000000053
0x00007ff8c0bb2fd9: winhttp!WinHttpClientSession::AcquireSession+0x0000000000000049
0x00007ff8c0bb3e34: winhttp!WinHttpClientResolver::GetProxyForUrlImpl+0x00000000000001a8
0x00007ff8c0bb3a48: winhttp!WinHttpClientResolver::GetProxyForUrlEx+0x0000000000000054
0x00007ff8c0bb50d9: winhttp!WinHttpClientCompletion::StartProxyResolve+0x0000000000000049
0x00007ff8c0b6102d: winhttp!WxProxyManager::OnProcessGetProxyForUrl+0x000000000002668d
0x00007ff8c0bb665b: winhttp!WxProxyManager::GetProxyForUrlImpl+0x00000000000000ef
0x00007ff8c0bb6518: winhttp!WxProxyManager::GetProxyForUrlEx+0x0000000000000054
0x00007ff8c0bb7683: winhttp!WxGetProxyContext::StartProxyResolve+0x0000000000000043
0x00007ff8c0b711ed: winhttp!HTTP_USER_REQUEST::_CallGetProxyForUrl+0x000000000003478d
--------------------------------------
--------------------------------------
Handle = 0x000000000000cfa0 - OPEN
Thread ID = 0x0000000000002d74, Process ID = 0x0000000000004528

0x00007ff8c74c165a: ntdll!ZwCreateEvent+0x000000000000000a
0x00007ff8c47c2ce8: KERNELBASE!CreateEventW+0x0000000000000084
0x00007ff8c0b2ed38: winhttp!INTERNET_SESSION_HANDLE_OBJECT::INTERNET_SESSION_HANDLE_OBJECT+0x00000000000002be
0x00007ff8c0b2e963: winhttp!WinHttpOpen+0x00000000000001f3
0x00007ff8c0bb9afb: winhttp!CreateWinHttpSession+0x0000000000000053
0x00007ff8c0bb2fd9: winhttp!WinHttpClientSession::AcquireSession+0x0000000000000049
0x00007ff8c0bb3e34: winhttp!WinHttpClientResolver::GetProxyForUrlImpl+0x00000000000001a8
0x00007ff8c0bb3a48: winhttp!WinHttpClientResolver::GetProxyForUrlEx+0x0000000000000054
0x00007ff8c0bb50d9: winhttp!WinHttpClientCompletion::StartProxyResolve+0x0000000000000049
0x00007ff8c0b6102d: winhttp!WxProxyManager::OnProcessGetProxyForUrl+0x000000000002668d
0x00007ff8c0bb665b: winhttp!WxProxyManager::GetProxyForUrlImpl+0x00000000000000ef
0x00007ff8c0bb6518: winhttp!WxProxyManager::GetProxyForUrlEx+0x0000000000000054
0x00007ff8c0bb7683: winhttp!WxGetProxyContext::StartProxyResolve+0x0000000000000043
0x00007ff8c0b711ed: winhttp!HTTP_USER_REQUEST::_CallGetProxyForUrl+0x000000000003478d
--------------------------------------
~~~

**Comment by davidsh** [comment]
This should be fixed now with #3096.

**Comment by davidsh** [comment]
@KKhurin The fix for the memory leaks is checked into the sources for System.Net.Http in the CoreFx repro.  

The fix is also available in the latest System.Net.Http/System.Net.Http.WinHttpHandler package version v4.0.1-beta-23310 from the CoreFx dev feed: https://www.myget.org/gallery/dotnet-core.
